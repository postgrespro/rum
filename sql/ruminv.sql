CREATE TABLE test_invrum(q tsquery);

INSERT INTO test_invrum VALUES ('a|b'::tsquery);
INSERT INTO test_invrum VALUES ('a&b'::tsquery);
INSERT INTO test_invrum VALUES ('!(a|b)'::tsquery);
INSERT INTO test_invrum VALUES ('!(a&b)'::tsquery);
INSERT INTO test_invrum VALUES ('!a|b'::tsquery);
INSERT INTO test_invrum VALUES ('a&!b'::tsquery);
INSERT INTO test_invrum VALUES ('(a|b)&c'::tsquery);
INSERT INTO test_invrum VALUES ('(!(a|b))&c'::tsquery);
INSERT INTO test_invrum VALUES ('(a|b)&(c|d)'::tsquery);
INSERT INTO test_invrum VALUES ('!a'::tsquery);
INSERT INTO test_invrum VALUES ('a&!(b&c)'::tsquery);
INSERT INTO test_invrum VALUES ('(a|a1|a2|a3|a4|a5)&(b|b1|b2|b3|b4|b5|b6)&!(c|c1|c2|c3)'::tsquery);

SELECT * FROM test_invrum WHERE q @@ ''::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'a'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'b'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'a b'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'c'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'a c'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'b c'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'a b c'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'd'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'a d'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'b d'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'a b d'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'c d'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'a c d'::tsvector;

CREATE INDEX test_invrum_idx ON test_invrum USING rum(q);
SET enable_seqscan = OFF;

SELECT * FROM test_invrum WHERE q @@ ''::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'a'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'b'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'a b'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'c'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'a c'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'b c'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'a b c'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'd'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'a d'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'b d'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'a b d'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'c d'::tsvector;
SELECT * FROM test_invrum WHERE q @@ 'a c d'::tsvector;

INSERT INTO test_invrum VALUES ('a:*'::tsquery);
INSERT INTO test_invrum VALUES ('a <-> b'::tsquery);
