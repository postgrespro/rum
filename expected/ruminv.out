CREATE TABLE test_invrum(q tsquery);
INSERT INTO test_invrum VALUES ('a|b'::tsquery);
INSERT INTO test_invrum VALUES ('a&b'::tsquery);
INSERT INTO test_invrum VALUES ('!(a|b)'::tsquery);
INSERT INTO test_invrum VALUES ('!(a&b)'::tsquery);
INSERT INTO test_invrum VALUES ('!a|b'::tsquery);
INSERT INTO test_invrum VALUES ('a&!b'::tsquery);
INSERT INTO test_invrum VALUES ('(a|b)&c'::tsquery);
INSERT INTO test_invrum VALUES ('(!(a|b))&c'::tsquery);
INSERT INTO test_invrum VALUES ('(a|b)&(c|d)'::tsquery);
INSERT INTO test_invrum VALUES ('!a'::tsquery);
INSERT INTO test_invrum VALUES ('a&!(b&c)'::tsquery);
INSERT INTO test_invrum VALUES ('(a|a1|a2|a3|a4|a5)&(b|b1|b2|b3|b4|b5|b6)&!(c|c1|c2|c3)'::tsquery);
SELECT * FROM test_invrum WHERE q @@ ''::tsvector;
       q        
----------------
 !( 'a' | 'b' )
 !( 'a' & 'b' )
 !'a' | 'b'
 !'a'
(4 rows)

SELECT * FROM test_invrum WHERE q @@ 'a'::tsvector;
          q           
----------------------
 'a' | 'b'
 !( 'a' & 'b' )
 'a' & !'b'
 'a' & !( 'b' & 'c' )
(4 rows)

SELECT * FROM test_invrum WHERE q @@ 'b'::tsvector;
       q        
----------------
 'a' | 'b'
 !( 'a' & 'b' )
 !'a' | 'b'
 !'a'
(4 rows)

SELECT * FROM test_invrum WHERE q @@ 'a b'::tsvector;
                                                               q                                                                
--------------------------------------------------------------------------------------------------------------------------------
 'a' | 'b'
 'a' & 'b'
 !'a' | 'b'
 'a' & !( 'b' & 'c' )
 ( 'a' | 'a1' | 'a2' | 'a3' | 'a4' | 'a5' ) & ( 'b' | 'b1' | 'b2' | 'b3' | 'b4' | 'b5' | 'b6' ) & !( 'c' | 'c1' | 'c2' | 'c3' )
(5 rows)

SELECT * FROM test_invrum WHERE q @@ 'c'::tsvector;
          q           
----------------------
 !( 'a' | 'b' )
 !( 'a' & 'b' )
 !'a' | 'b'
 !( 'a' | 'b' ) & 'c'
 !'a'
(5 rows)

SELECT * FROM test_invrum WHERE q @@ 'a c'::tsvector;
               q               
-------------------------------
 'a' | 'b'
 !( 'a' & 'b' )
 'a' & !'b'
 ( 'a' | 'b' ) & 'c'
 ( 'a' | 'b' ) & ( 'c' | 'd' )
 'a' & !( 'b' & 'c' )
(6 rows)

SELECT * FROM test_invrum WHERE q @@ 'b c'::tsvector;
               q               
-------------------------------
 'a' | 'b'
 !( 'a' & 'b' )
 !'a' | 'b'
 ( 'a' | 'b' ) & 'c'
 ( 'a' | 'b' ) & ( 'c' | 'd' )
 !'a'
(6 rows)

SELECT * FROM test_invrum WHERE q @@ 'a b c'::tsvector;
               q               
-------------------------------
 'a' | 'b'
 'a' & 'b'
 !'a' | 'b'
 ( 'a' | 'b' ) & 'c'
 ( 'a' | 'b' ) & ( 'c' | 'd' )
(5 rows)

SELECT * FROM test_invrum WHERE q @@ 'd'::tsvector;
       q        
----------------
 !( 'a' | 'b' )
 !( 'a' & 'b' )
 !'a' | 'b'
 !'a'
(4 rows)

SELECT * FROM test_invrum WHERE q @@ 'a d'::tsvector;
               q               
-------------------------------
 'a' | 'b'
 !( 'a' & 'b' )
 'a' & !'b'
 ( 'a' | 'b' ) & ( 'c' | 'd' )
 'a' & !( 'b' & 'c' )
(5 rows)

SELECT * FROM test_invrum WHERE q @@ 'b d'::tsvector;
               q               
-------------------------------
 'a' | 'b'
 !( 'a' & 'b' )
 !'a' | 'b'
 ( 'a' | 'b' ) & ( 'c' | 'd' )
 !'a'
(5 rows)

SELECT * FROM test_invrum WHERE q @@ 'a b d'::tsvector;
                                                               q                                                                
--------------------------------------------------------------------------------------------------------------------------------
 'a' | 'b'
 'a' & 'b'
 !'a' | 'b'
 ( 'a' | 'b' ) & ( 'c' | 'd' )
 'a' & !( 'b' & 'c' )
 ( 'a' | 'a1' | 'a2' | 'a3' | 'a4' | 'a5' ) & ( 'b' | 'b1' | 'b2' | 'b3' | 'b4' | 'b5' | 'b6' ) & !( 'c' | 'c1' | 'c2' | 'c3' )
(6 rows)

SELECT * FROM test_invrum WHERE q @@ 'c d'::tsvector;
          q           
----------------------
 !( 'a' | 'b' )
 !( 'a' & 'b' )
 !'a' | 'b'
 !( 'a' | 'b' ) & 'c'
 !'a'
(5 rows)

SELECT * FROM test_invrum WHERE q @@ 'a c d'::tsvector;
               q               
-------------------------------
 'a' | 'b'
 !( 'a' & 'b' )
 'a' & !'b'
 ( 'a' | 'b' ) & 'c'
 ( 'a' | 'b' ) & ( 'c' | 'd' )
 'a' & !( 'b' & 'c' )
(6 rows)

CREATE INDEX test_invrum_idx ON test_invrum USING rum(q);
SET enable_seqscan = OFF;
SELECT * FROM test_invrum WHERE q @@ ''::tsvector;
 q 
---
(0 rows)

SELECT * FROM test_invrum WHERE q @@ 'a'::tsvector;
          q           
----------------------
 'a' | 'b'
 !( 'a' & 'b' )
 'a' & !'b'
 'a' & !( 'b' & 'c' )
(4 rows)

SELECT * FROM test_invrum WHERE q @@ 'b'::tsvector;
       q        
----------------
 'a' | 'b'
 !( 'a' & 'b' )
 !'a' | 'b'
 !'a'
(4 rows)

SELECT * FROM test_invrum WHERE q @@ 'a b'::tsvector;
                                                               q                                                                
--------------------------------------------------------------------------------------------------------------------------------
 'a' | 'b'
 'a' & 'b'
 !'a' | 'b'
 ( 'a' | 'a1' | 'a2' | 'a3' | 'a4' | 'a5' ) & ( 'b' | 'b1' | 'b2' | 'b3' | 'b4' | 'b5' | 'b6' ) & !( 'c' | 'c1' | 'c2' | 'c3' )
(4 rows)

SELECT * FROM test_invrum WHERE q @@ 'c'::tsvector;
          q           
----------------------
 !( 'a' | 'b' )
 !( 'a' & 'b' )
 !'a' | 'b'
 !( 'a' | 'b' ) & 'c'
 !'a'
(5 rows)

SELECT * FROM test_invrum WHERE q @@ 'a c'::tsvector;
               q               
-------------------------------
 'a' | 'b'
 !( 'a' & 'b' )
 'a' & !'b'
 ( 'a' | 'b' ) & 'c'
 ( 'a' | 'b' ) & ( 'c' | 'd' )
(5 rows)

SELECT * FROM test_invrum WHERE q @@ 'b c'::tsvector;
               q               
-------------------------------
 'a' | 'b'
 !( 'a' & 'b' )
 !'a' | 'b'
 ( 'a' | 'b' ) & 'c'
 ( 'a' | 'b' ) & ( 'c' | 'd' )
 !'a'
(6 rows)

SELECT * FROM test_invrum WHERE q @@ 'a b c'::tsvector;
               q               
-------------------------------
 'a' | 'b'
 'a' & 'b'
 !'a' | 'b'
 ( 'a' | 'b' ) & 'c'
 ( 'a' | 'b' ) & ( 'c' | 'd' )
(5 rows)

SELECT * FROM test_invrum WHERE q @@ 'd'::tsvector;
       q        
----------------
 !( 'a' | 'b' )
 !( 'a' & 'b' )
 !'a' | 'b'
 !'a'
(4 rows)

SELECT * FROM test_invrum WHERE q @@ 'a d'::tsvector;
               q               
-------------------------------
 'a' | 'b'
 !( 'a' & 'b' )
 'a' & !'b'
 ( 'a' | 'b' ) & ( 'c' | 'd' )
 'a' & !( 'b' & 'c' )
(5 rows)

SELECT * FROM test_invrum WHERE q @@ 'b d'::tsvector;
               q               
-------------------------------
 'a' | 'b'
 !( 'a' & 'b' )
 !'a' | 'b'
 ( 'a' | 'b' ) & ( 'c' | 'd' )
 !'a'
(5 rows)

SELECT * FROM test_invrum WHERE q @@ 'a b d'::tsvector;
                                                               q                                                                
--------------------------------------------------------------------------------------------------------------------------------
 'a' | 'b'
 'a' & 'b'
 !'a' | 'b'
 ( 'a' | 'b' ) & ( 'c' | 'd' )
 ( 'a' | 'a1' | 'a2' | 'a3' | 'a4' | 'a5' ) & ( 'b' | 'b1' | 'b2' | 'b3' | 'b4' | 'b5' | 'b6' ) & !( 'c' | 'c1' | 'c2' | 'c3' )
(5 rows)

SELECT * FROM test_invrum WHERE q @@ 'c d'::tsvector;
          q           
----------------------
 !( 'a' | 'b' )
 !( 'a' & 'b' )
 !'a' | 'b'
 !( 'a' | 'b' ) & 'c'
 !'a'
(5 rows)

SELECT * FROM test_invrum WHERE q @@ 'a c d'::tsvector;
               q               
-------------------------------
 'a' | 'b'
 !( 'a' & 'b' )
 'a' & !'b'
 ( 'a' | 'b' ) & 'c'
 ( 'a' | 'b' ) & ( 'c' | 'd' )
(5 rows)

INSERT INTO test_invrum VALUES ('a:*'::tsquery);
ERROR:  Indexing of prefix tsqueries isn't supported yet
INSERT INTO test_invrum VALUES ('a <-> b'::tsquery);
ERROR:  Indexing of phrase tsqueries isn't supported yet
